File: src/main/java/cad97/spawnercraft/init/SpawnerCraftRecipes.java
package cad97.spawnercraft.init;

import cad97.spawnercraft.handler.ConfigHandler;
import cad97.spawnercraft.utility.LogHelper;
import cad97.spawnercraft.utility.NBTPreservingShapedRecipe;
import net.minecraft.init.Blocks;
import net.minecraft.init.Items;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.Ingredient;
import net.minecraft.util.NonNullList;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fml.common.registry.ForgeRegistries;
import net.minecraftforge.fml.common.registry.GameRegistry;

public class SpawnerCraftRecipes {
    public static void registerRecipes() {
        Ingredient essence = Ingredient.fromItem(SpawnerCraftItems.MOB_ESSENCE);
        ForgeRegistries.RECIPES.register(new NBTPreservingShapedRecipe(
                "spawnercraft",
                2, 2,
                NonNullList.withSize(4, essence),
                new ItemStack(SpawnerCraftItems.MOB_AGGLOMERATION)
        ).setRegistryName("craft_mob_agglomeration"));
        Ingredient agglomeration = Ingredient.fromItem(SpawnerCraftItems.MOB_AGGLOMERATION);
        ForgeRegistries.RECIPES.register(new NBTPreservingShapedRecipe(
                "spawnercraft",
                2, 2,
                NonNullList.withSize(4, agglomeration),
                new ItemStack(SpawnerCraftItems.MOB_SPIRIT)
        ).setRegistryName("craft_mob_spirit"));
        if (ConfigHandler.spawnerCraftable) {
            GameRegistry.addShapedRecipe(
                    new ResourceLocation("spawnercraft", "craft_mob_cage"),
                    null,
                    new ItemStack(SpawnerCraftBlocks.MOB_CAGE),
                    "III",
                    "I I",
                    "III",
                    'I', new ItemStack(Blocks.IRON_BARS)
            );
        }
        GameRegistry.addShapedRecipe(
                new ResourceLocation("spawnercraft", "craft_mob_rod"),
                null,
                new ItemStack(SpawnerCraftItems.MOB_ROD),
                "F",
                "S",
                'F', new ItemStack(Items.FISHING_ROD),
                'S', new ItemStack(SpawnerCraftBlocks.MOB_CAGE)
        );
        LogHelper.logInfo("Recipes registered.");
    }
}

File: src/main/kotlin/cad97/spawnercraftkt/init/SpawnerCraftRecipes.kt
package cad97.spawnercraftkt.init

import cad97.spawnercraftkt.Config
import cad97.spawnercraftkt.SpawnerCraft
import cad97.spawnercraftkt.block.BlockMobCage
import net.minecraft.init.Blocks
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.IRecipe
import net.minecraft.item.crafting.Ingredient
import net.minecraft.item.crafting.ShapedRecipes
import net.minecraft.util.NonNullList
import net.minecraftforge.event.RegistryEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent

@Mod.EventBusSubscriber
object SpawnerCraftRecipes {
    @JvmStatic
    @SubscribeEvent
    fun registerRecipes(event: RegistryEvent.Register<IRecipe>) {
        if (Config.spawnerCraftable) {
            val ironBars = Ingredient.fromItem(Item.getItemFromBlock(Blocks.IRON_BARS))
            event.registry.register(ShapedRecipes(
                    "",
                    3, 3,
                    NonNullList.from(null,
                            ironBars, ironBars, ironBars,
                            ironBars, Ingredient.EMPTY, ironBars,
                            ironBars, ironBars, ironBars
                    ),
                    ItemStack(SpawnerCraftBlocks.mob_cage)
            ).setRegistryName(SpawnerCraft.modid, BlockMobCage.id))
        }
        SpawnerCraft.logger.info("Recipes registered.")
    }
}
